src/task.c

```diff
@@ -747,25 +686,116 @@ task_group_perform_one(task_group *group)
 platform_status
 task_perform_one(task_system *ts)
 {
-   platform_assert(!ts->use_bg_threads);
    platform_status rc = STATUS_OK;
    for (task_type type = TASK_TYPE_FIRST; type != NUM_TASK_TYPES; type++) {
-      rc = task_group_perform_one(&ts->group[type]);
+      rc = task_group_perform_one(&ts->group[type], 0);
+      /* STATUS_TIMEDOUT means no task was waiting. */
+      if (STATUS_IS_NE(rc, STATUS_TIMEDOUT)) {
+         return rc;
+      }
+   }
+   return rc;
+}
+
+platform_status
+task_perform_one_if_needed(task_system *ts)
+{
+   platform_status rc = STATUS_OK;
+   for (task_type type = TASK_TYPE_FIRST; type != NUM_TASK_TYPES; type++) {
+      /* Perform a task only if there are more waiting tasks than bg
```

Comment:
Maybe move this hidden common as prolog comment block, formatted like it's done for `task_perform_one()`.